File: docs/check.js
// 音階にできない音を弾く。
// ../src/voice_scale/check.py と同じ閾値。

export const MIN_PEAK = 0.01;
export const MIN_SEC = 0.15;
export const MIN_VOICED = 0.5;
export const MAX_SPREAD_CENTS = 150;

// 教室で何度も弾かれると子どもが萎縮する。原理的に無理なものだけ確実に止め、
// 多少音程が動く程度は通す。150セントはそのために緩めに取ってある。
export const MESSAGES = {
  quiet: 'もっと ちかくで こえを だしてね',
  short: 'もうすこし ながく こえを だしてね',
  unpitched: '「あー」や「にゃー」のように こえを のばしてね',
  unstable: 'おなじ たかさで こえを だしてね',
};

export function peakOf(x) {
  let peak = 0;
  for (let i = 0; i < x.length; i += 1) {
    const v = Math.abs(x[i]);
    if (v > peak) peak = v;
  }
  return peak;
}

// 弾く理由のキーを並べて返す。空なら音階にできる。
export function judge(x, sr, pitch) {
  const reasons = [];
  if (peakOf(x) < MIN_PEAK) reasons.push('quiet');
  if (x.length / sr < MIN_SEC) reasons.push('short');
  if (pitch.voicedRatio < MIN_VOICED || !Number.isFinite(pitch.f0)) reasons.push('unpitched');
  else if (pitch.spreadCents > MAX_SPREAD_CENTS) reasons.push('unstable');
  return reasons;
}

File: src/voice_scale/check.py
"""音階にできない音を弾く。

docs/check.js と対になる。

無声音やノイズは声帯が震えていないので波が規則正しくならず、
原理的に音程が存在しない。
"""

from __future__ import annotations

import numpy as np

from voice_scale.pitch import Pitch

MIN_PEAK = 0.01  # これ未満は小さすぎる
MIN_SEC = 0.15  # これ未満は短すぎる
MIN_VOICED = 0.50  # 有声窓がこの割合を下回れば音程なし
MAX_SPREAD_CENTS = 150.0  # 中央値からのばらつきの上限

# 教室で何度も弾かれると子どもが萎縮する。原理的に無理なものだけ確実に止め、
# 多少音程が動く程度は通す。閾値150セントはそのために緩めに取ってある。

MESSAGES = {
    "quiet": "もっと近くで声を出してね",
    "short": "もう少し長く声を出してね",
    "unpitched": "「あー」や「にゃー」のように声をのばしてね",
    "unstable": "同じ高さで声を出してね",
}


def judge(x: np.ndarray, sr: int, pitch: Pitch) -> list[str]:
    """弾く理由のキーを並べて返す。空なら音階にできる。"""
    reasons: list[str] = []
    x = np.asarray(x, dtype=np.float64)
    peak = float(np.max(np.abs(x))) if len(x) else 0.0

    if peak < MIN_PEAK:
        reasons.append("quiet")
    if len(x) / sr < MIN_SEC:
        reasons.append("short")
    if pitch.voiced_ratio < MIN_VOICED or not np.isfinite(pitch.f0):
        reasons.append("unpitched")
    elif pitch.spread_cents > MAX_SPREAD_CENTS:
        reasons.append("unstable")
    return reasons
